hypothesis-python/src/hypothesis/searchstrategy/types.py

```diff
@@ -177,6 +177,21 @@ def from_typing_type(thing):
         # such as Container
         for t in (typing.KeysView, typing.ValuesView, typing.ItemsView):
             mapping.pop(t, None)
+    if len(mapping) > 1:
+        # issubclass treats bytestring as a kind of sequence, which it is,
+        # but treating it as such breaks everything else when it is presumed
+        # to be a generic sequence or container that could hold any item.
+        # Except for sequences of integers, or unions which include integer!
+        # See https://github.com/HypothesisWorks/hypothesis/issues/2257
+        elem_type = (getattr(thing, "__args__", None) or {0: str})[0]
```

Comment:
Why do we default to assuming `elem_type` is str?

In general, I would find this code easier to follow if the comment explained what it is doing instead of just what problem it is solving. I *think* what it is doing is removing `ByteString` unless the element type allows integers, but I'm not clear on this.

Reply:
It could be anything that's not `int`, so perhaps `"not int"` would be clearer!

And yep, I'll expand that comment - it drops `ByteString` unless the element type is `int` or a union including `int`.

Reply:
Wouldn't it also make sense to generate a `ByteString` if asked for a `Sequence[Object]`?

This seems like a more general problem rather than anything to do with `ByteString` - you only want to generate things of type `Sequence[A]` in response to a request for things of type `Sequence[B]` if `A` is a subtype (not subclass!) of `B`. Isn't that general compatibility rule really what you should be testing for here?

Reply:
> you only want to generate things of type `Sequence[A]` in response to a request for things of type `Sequence[B]` if `A` is a subtype (not subclass!) of `B`

That's exactly what we do *if the sequence subtype we generate is generic*; but having the non-generic type ByteString registered in our mapping of known types messes this up since it inherits from the generic `Sequence[int]` (but the runtime implementation doesn't track the element type) and it's also defined in the typing module and thus subject to special handling.

This is very messy, but that's basically just how `from_type` goes with generics :sob: